src/gafaelfawr/handlers/oidc.py

```diff
@@ -0,0 +1,229 @@
+        return_url = build_return_url(
+            parsed_redirect_uri,
+            state=state,
+            error=e.error,
+            error_description=str(e),
+        )
+        raise web.HTTPFound(return_url)
+
+    # Get an authorization code and return it.
+    code = await oidc_server.issue_code(
+        client_id, redirect_uri, session.handle
+    )
+    return_url = build_return_url(
+        parsed_redirect_uri, state=state, code=code.encode()
+    )
+    context.logger.info("Returned OpenID Connect authorization code")
+    raise web.HTTPFound(return_url)
+
+
+def build_return_url(
+    redirect_uri: ParseResult, **params: Optional[str]
+) -> str:
+    """Construct a return URL for a redirect.
+
+    Parameters
+    ----------
+    redirect_uri : `urllib.parse.ParseResult`
+        The parsed return URI from the client.
+    **params : `str` or `None`
+        Additional parameters to add to that URI to create the return URL.
+        Any parameters set to `None` will be ignored.
+
+    Returns
+    -------
+    return_url : `str`
+        The return URL to which the user should be redirected.
+    """
+    query = parse_qsl(redirect_uri.query) if redirect_uri.query else []
+    query.extend(((k, v) for (k, v) in params.items() if v is not None))
+    return_url = redirect_uri._replace(query=urlencode(query))
+    return return_url.geturl()
+
+
+def get_form_value(data: MultiDictProxy, key: str) -> Optional[str]:
+    """Retrieve a string from form data.
+
+    This handles encoding issues and returns `None` if one of the fields is
+    unexpectedly a file upload.
+
+    Parameters
+    ----------
+    data : `multidict.MultiDictProxy`
+        The form data.
+    key : `str`
+        The field to extract.
+
+    Returns
+    -------
+    value : `str` or `None`
+        The value, or `None` if this field wasn't present or if it was a file
+        upload.
+    """
+    value = data.get(key)
+    if not value:
+        return None
+    elif isinstance(value, str):
+        return value
+    elif isinstance(value, bytes):
+        return value.decode()
+    else:
+        return None
+
+
+@routes.post("/auth/openid/token")
+async def post_token(request: web.Request) -> web.Response:
+    """Redeem an authorization code for a token.
+
+    Parameters
+    ----------
+    request : `aiohttp.web.Request`
+        The incoming request.
+
+    Returns
+    -------
+    response : `aiohttp.web.Response`
+        The response.
+    """
+    context = RequestContext.from_request(request)
+    data = await request.post()
+    grant_type = get_form_value(data, "grant_type")
+    client_id = get_form_value(data, "client_id")
+    client_secret = get_form_value(data, "client_secret")
+    code = get_form_value(data, "code")
+    redirect_uri = get_form_value(data, "redirect_uri")
+
+    # Check the request parameters.
+    if not grant_type or not client_id or not code or not redirect_uri:
```

Comment:
if not all([grant_type, client_id, code, redirect_uri])?

Reply:
Oh, good idea. I hadn't thought of that trick.

Reply:
Sadly, mypy hadn't thought of that trick either, and it causes type checking to not work because it doesn't realize that these variables can no longer be `None`.